### PR TITLE
[feature] sift: update default values, add new filtering and add dsp-sift variation

### PR DESCRIPTION
**Pull request by fabiencastan**
## Description

Feature point quality has a critical impact on SfM and Panorama pipelines.
This PR re-evaluate the default parameters and add new filtering strategies to be less impacted by the peak threshold in standard sift. This parameter is critical in particular on HDR images.

## Features list

- [X] Update default sift thresholds
- [X] Add new feature points filtering
- [X] Add dsp-sift variation

## Implementation remarks

 - [X] Update vlfeat to 0.9.21
 - [X] vlfeat: build as shared library and remove init/uninit on dll load
 - [X] vlfeat: remove vlcov internal buffer to allow multi-threading (use an explicit parameter instead)

### CODE REVIEW

File: src/aliceVision/feature/ImageDescriber.hpp [hunk 1]
```diff
@@ -31,35 +31,113 @@ enum class EImageDescriberPreset
...
+};
+
+inline std::string EFeatureQuality_information()
+{
+    return "Feature extraction contains a compromize between speed and precision:\n"
+           "* LOW: Very quick results.\n"
+           "* MEDIUM: Quick results.\n"
+           "* NORMAL: Default feature quality.\n"
+           "* HIGH: Improved quality over performances.\n"
+           "* ULTRA: Perfect quality at the expense of high computational cost.\n";
```
Comment (servantf) [inline src/aliceVision/feature/ImageDescriber.hpp:68]: Perfect -> Best ? :p

File: src/aliceVision/feature/sift/ImageDescriber_DSPSIFT_vlfeat.cpp [hunk 2]
```diff
@@ -0,0 +1,393 @@
...
+                {
+                    const double dspScale = dspMinScale + s * dspScaleStep;
+
+                    VlFrameOrientedEllipse frameAtScale = inFeat.frame;
+                    frameAtScale.a11 *= dspScale;
+                    frameAtScale.a12 *= dspScale;
+                    frameAtScale.a21 *= dspScale;
+                    frameAtScale.a22 *= dspScale;
+
+                    vl_covdet_extract_patch_for_frame(covdet.get(), patch.data(), kPatchResolution, &internalBuffer,
```
Comment (servantf) [inline src/aliceVision/feature/sift/ImageDescriber_DSPSIFT_vlfeat.cpp:339]: To me, this is extracting a sub image at a given location at an interpolated scale.

I assume the scale interpolation is based on the existing, pre-computed scales. So this is bounded by the parameters of the sift detector. As such, shouldn't the scales be computed to make sure that for each keypoint, we are able to  get a range of scales large enough ?

**Review by servantf — DISMISSED** [review]

File: src/aliceVision/feature/ImageDescriber.hpp (hunk 1 above)
Comment (simogasp) [reply src/aliceVision/feature/ImageDescriber.hpp:68]: Highest?

File: src/aliceVision/feature/ImageDescriber.hpp [hunk 3]
```diff
@@ -31,35 +31,115 @@ enum class EImageDescriberPreset
...
+    GridSortOctaveSteps,
+    NonExtremaFiltering
+};
+
+inline std::string EFeatureConstrastFiltering_information()
+{
+    return "Contrast filtering method to ignore features with too low contrast that can be consided as noise:\n"
+           "* Static: Fixed threshold.\n"
+           "* AdaptiveToMedianVariance: Based on image content analysis.\n"
+           "* NoFiltering: Disable contrast filtering.\n";
```
Comment (simogasp) [inline src/aliceVision/feature/ImageDescriber.hpp:97]: what about the others?

File: src/aliceVision/feature/ImageDescriber.hpp [hunk 4]
```diff
@@ -31,35 +31,115 @@ enum class EImageDescriberPreset
...
- * @brief It read a describer preset from a stream. 
- * @param[in] in the stream from which the preset is read.
- * @param[out] p the preset read from the stream.
- * @return the modified stream without the read preset.
+ * @brief The method used to filter out features with too low constrast (that can be considered as noise).
  */
-std::istream& operator>>(std::istream& in, EImageDescriberPreset& p);
+enum class EFeatureConstrastFiltering
+{
+    Static = 0,
```
Comment (simogasp) [inline src/aliceVision/feature/ImageDescriber.hpp:82]:
```suggestion
    /// Use a fixed threshold for all the pixels
    Static = 0,
```
etc...

File: src/aliceVision/feature/ImageDescriber.hpp [hunk 5]
```diff
@@ -31,35 +31,115 @@ enum class EImageDescriberPreset
...
+
+EFeatureConstrastFiltering EFeatureConstrastFiltering_stringToEnum(const std::string& v);
+std::string EFeatureConstrastFiltering_enumToString(const EFeatureConstrastFiltering v);
+std::ostream& operator<<(std::ostream& os, EFeatureConstrastFiltering v);
+std::istream& operator>>(std::istream& in, EFeatureConstrastFiltering& v);
+
+
+struct ConfigurationPreset
+{
+    EImageDescriberPreset descPreset = EImageDescriberPreset::NORMAL;
```
Comment (simogasp) [inline src/aliceVision/feature/ImageDescriber.hpp:108]: prefer `{}` for initialization

File: src/aliceVision/feature/imageDescriberCommon.cpp [hunk 6]
```diff
@@ -75,6 +78,9 @@ EImageDescriberType EImageDescriberType_stringToEnum(const std::string& imageDes
   if(type == "sift")          return EImageDescriberType::SIFT;
   if(type == "sift_float")    return EImageDescriberType::SIFT_FLOAT;
   if(type == "sift_upright")  return EImageDescriberType::SIFT_UPRIGHT;
+
+  if(type == "dspsift")       return EImageDescriberType::DSPSIFT;
+
```
Comment (simogasp) [inline src/aliceVision/feature/imageDescriberCommon.cpp:83]:
```suggestion
  if(type == "dspsift")       return EImageDescriberType::DSPSIFT;
```

File: src/aliceVision/feature/imageDescriberCommon.hpp [hunk 7]
```diff
@@ -23,6 +23,9 @@ enum class EImageDescriberType: unsigned char
   , SIFT = 10
   , SIFT_FLOAT = 11
   , SIFT_UPRIGHT = 12
+
+  , DSPSIFT = 13
+
```
Comment (simogasp) [inline src/aliceVision/feature/imageDescriberCommon.hpp:28]:
```suggestion
  , DSPSIFT = 13
```

File: src/aliceVision/feature/imageStats.hpp [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+namespace aliceVision {
+namespace feature {
+
+/**
+ * @brief Compute Contrast Factor
+ * @param[in] image Input image for the given octave
+ * @param[in] percentile
+ * @return contrastFactor
+ */
+inline float computeAutomaticContrastFactor(const image::Image<float>& image, const float percentile)
```
Comment (simogasp) [inline src/aliceVision/feature/imageStats.hpp:22]: `alicevision_feature` is not a header only lib anymore, we can split the code, even if it is just one function

File: src/aliceVision/feature/imageStats.hpp [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+#pragma once
+
+#include <aliceVision/image/all.hpp>
+
+
+namespace aliceVision {
+namespace feature {
+
+/**
+ * @brief Compute Contrast Factor
```
Comment (simogasp) [inline src/aliceVision/feature/imageStats.hpp:17]: There are different definitions of Contrast Factor, would be better to explain briefly what is compute here

File: src/aliceVision/matching/svgVisualization.cpp [hunk 10]
```diff
@@ -25,6 +25,8 @@ std::string describerTypeColor(feature::EImageDescriberType descType )
     case feature::EImageDescriberType::SIFT:           return "yellow";
     case feature::EImageDescriberType::SIFT_FLOAT:     return "yellow";
     case feature::EImageDescriberType::SIFT_UPRIGHT:   return "yellow";
+    case feature::EImageDescriberType::DSPSIFT:        return "yellow";
+
```
Comment (simogasp) [inline src/aliceVision/matching/svgVisualization.cpp:29]:
```suggestion
    case feature::EImageDescriberType::DSPSIFT:        return "yellow";
```

File: src/aliceVision/feature/sift/SIFT.cpp [hunk 11]
```diff
@@ -11,18 +11,104 @@ namespace feature {
...
 std::size_t getMemoryConsumptionVLFeat(std::size_t width, std::size_t height, const SiftParams& params)
 {
   double scaleFactor = 1.0;
-  if(params._firstOctave > 0)
-      scaleFactor = 1.0 / std::pow(2.0, params._firstOctave);
-  else if(params._firstOctave < 0)
-      scaleFactor = std::pow(2.0, std::abs(params._firstOctave));
+
+  // if image resolution is low, increase resolution for extraction
+  const int firstOctave = params._firstOctave - (width*height < 5000*4000 ? 1 : 0);
```
Comment (simogasp) [inline src/aliceVision/feature/sift/SIFT.cpp:100]: would be better to refactor this as the condition is used several times in the code, better to have it in only onw place for maintenability

**Review by simogasp — CHANGES_REQUESTED** [review]

File: src/aliceVision/matching/svgVisualization.cpp (hunk 10 above)
Comment (fabiencastan) [reply src/aliceVision/matching/svgVisualization.cpp:29]: I do not see the change.

**Pull request merged**